[WIP] SA2.0, model/data-access edits, unit testing

jdavcs opened the pull request:
Simple model and data-access unit testing. Db created and loaded once per session; cleared after each test. Dependent model instances created recursively as needed, with provided arguments or reasonable defaults.
Draft.
Builds on #17180.



## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
[TMP] Combined commits from #17180

Upgrade SQLAlchemy to 2.0

This conflicts with dependency requirements for sqlalchemy-graphene
(used only in toolshed, new WIP client)

Remove RemovedIn20Warning from config

This does not exist in SQLAlchemy 2.0

Update import path for DeclarativeMeta

Move declaration of injected attrs into constructor
Remove unused import

For context: https://github.com/galaxyproject/galaxy/pull/14717/files#r1486979280

Also, remove model attr type hints that conflict with SA2.0

Apply Mapped/mapped_column to model definitions

Included models: galaxy, tool shed, tool shed install
Column types:
DateTime
Integer
Boolan
Unicode
String (Text/TEXT/TrimmedString/VARCHAR)
UUID
Numeric

NOTE on typing of nullability: db schema != python app

- Mapped[datetime] specifies correct type for the python app;
- nullable=True specifies correct mapping to the db schema (that's what
  the CREATE TABLE sql statement will reflect).

mapped_column.nullable takes precedence over typing annotation of
Mapped. So, if we have:

foo: Mapped[str] = mapped_column(String, nullable=True)

- that means that the foo db field will allow NULL, but the python app
  will not allow foo = None. And vice-versa:

bar: Mapped[Optional[str]] = mapped_column(String, nullable=False)

- the bar db field is NOT NULL, but bar = None is OK.

This might need to be applied to other column definitions, but for now
this addresses specific mypy errors.

Ref: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapped-column-derives-the-datatype-and-nullability-from-the-mapped-annotation

Add typing to JSON columns, fix related mypy errors

Columns:
MutableJSONType
JSONType
DoubleEncodedJsonType

TODO: I think we need a type alias for json-typed columns: bytes understand
iteration, but not access by key.

Use correct type hints to define common model attrs

Start applying Mapped to relationship definitions in the model

Remove column declaration from HasTags parent class

Fix SA2.0 error: wrap sql in text()

Fix SA2.0 error: pass bind to create_all

Fix SA2.0 error: use Row._mapping for keyed attribute access

Ref: https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#result-rows-act-like-named-tuples

Fix SA2.0 error: show password in url

SA 1.4: str(url) renders connection string with password
SA 2.0: str(url) renders connection string WITHOUT password
Solution: Use render_as_string(hide_password=False)

Fix SA2.0 error: use attribute_keyed_dict

Replaces attribute_mapped_collection (SA20)

Fix SA2.0 error: make select stmt a subquery

Rename varable to fix mypy

Fix SA2.0 error: explicitly use subquery() for select-from argument

Fix SA2.0 error: replase session.bind with session.get_bind()

Fix SA2.0 error: joinedload does not take str args

Fix use of table model attribute

- Use __table__ (SA attr) instead of table (galaxy attr) on mapped classes
- Drop .table and .table.c where redundant

Fix bug: fix HistoryAudit model

It is not a child of RepresentById becuase it does not and should not have an id attr.

Duplicating the __repr__ definition in the HistoryAudit class is a
temporary fix: a proper fix requires changing all models (id and
__repr__ should be split into 2 mixins): to be done in a follow-up PR.

Fix bug: check if template.fields is not null before iterating

Fix bug: call unique() on result, not select stmt

Fix bug: do not pass subquery to in_

Fix bug/typo: use select_from

Fix bug: if using alias on ORM entity, use __table__ as valid FromClause

Fix bug: HDAH model is not serializable (caught by mypy)

Fix typing error: migrations.base

Fix typing error: managers.secured

This fixed 58 mypy errors!

Fix typing error: session type

Fix typing error: use Session instead of scoped_session

No need to pass around scoped_session as arguments

Fix typing error: sharable

Fix SA2.0 error: sqlalchemy exceptions import; minor mypy fix

Mypy: type-ignore: this is never SessionlessContext

Mypy: use verbose assignment to help mypy

Mypy: add assert stmt

Mypy: add assert to ensure seesion is not None

Calling that method when a User obj is not attached to a session should
not happen.

Mypy: return 0 if no results

Mypy: type-ignore: scoped_session vs. install_model_session

We use the disctinction for DI.

Mypy: refactor to one-liner

Mypy: add assert stmts where we know session returns an object

Mypy: rename wfi_step > wfi_step_sq when it becomes a subquery

Job search refactor: factor out build_job_subquery

Job search refactor: build_stmt_for_hda

Job search refactor: build_stmt_for_ldda

Job search refactor: build_stmt_for_hdca

Job search refactor: build_stmt_for_dce

Job search refactor: rename query >> stmt

Mypy: add anno for Lists; type-ignore for HDAs

Note: type-ignore is due to imperative mapping of HDAs (and LDDAs). This
will be removed once we map those models declaratively

Mypy: managers.histories

Mypy: model.deferred

Mypy: arg passed to template can be None

Mypy: celery tasks

type ignore arg: we need to map DatasetInstance classes declaratively
for that to work correctly.

Mypy: type-ignore hda attr-defined error

Need to map declaratively to remove this

Convert visualization manager index query to SA Core

Mypy: session is not none

Mypy: type-ignore what requires more refactoring

Mypy: type-ignore hda, ldda attrs: need declarative mapping

Also, minor SA2.0 syntax fix

Mypy: type-ignores to handle late evaluation of relationship arguments

Mypy: type-ignore column property assignments (type is correct)

Mypy: typing errors, misc. fixes

Mypy: all statements are reachable

Mypy: need to map hda declaratively, then its parent is model.Base

Fix typing errors: sharable, secured

Fix package mypy errors

Fix SA2.0 error: celery task

1. In 2.0, when the statement contains "returning", the result type is
   ChunkedIteratorResult, which does not have the rowcount attr,
   becuase:
2. result.rowcount should not be used for statements containting the returning clause

Ref: https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.rowcount

Wrap call to ensure session is closed

Otherwise there's an idle transaction left in the database (+locks)

Ensure session is closed on TS Registry load

Same as prev. commit: otherwise db locks are left

Fix SA2.0 error: list arg to select; mypy

Use NullPool for sqlite engines

This restores the behavior under SQLAlchemy 1.4
(Note that we set the pool for sqlite only if it's not an in-memory db

Help mypy: job is never None

jdavcs committed:
security.get_npns_roles: test + factor out

jdavcs committed:
security.get_private_user_role: test + factor out

jdavcs committed:
webapps.galaxy.services.user.get_users_for_index: test + factor out

jmchilton commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1181,7 +1181,7 @@ def galaxy_url(self):
 
     def get_job(self) -> model.Job:
-        return self.sa_session.get(Job, self.job_id)
+        return self.sa_session.get(Job, self.job_id)  # type:ignore[return-value]
```
This is a lot of type ignores in this PR. This one feels of an easier variety....

Can we have a helper utility that does that and does a typing.cast to prevent the typing ignore here.

```
from typing import cast, TypeVar

T = TypeVar('T')

def load_from_session(session, clazz: T,  id: int) -> T:
    val = session.get(clazz, id)
    return cast(clazz, val)
```

I don't think we want a typing ignore every place we are doing a session.get.

It looks like a lot of these are scalars() which is untyped... maybe we could do a typed version of scalars that does the cast also? I am not sure if that makes sense but I would love to centralize the ignores if possible. Random application code shouldn't have to have these ignores in there I think.

It looks like a lot of these ignores are ``type:ignore[return-value]`` also - I would prefer setting a variable of the right type and ignoring the problem there and returning "the right thing" without a type ignore. Maybe that is silly - but to me that would feel like the hacks are more enclosed in their blackboxes.

jdavcs replied:
[This edit was part of the SA2.0 PR (#17180)] 

If I'm not mistaken, this type-ignore is there to ignore the None value that the `session.get` may return (by definition of `session.get`). The non-None value should be typed correctly (`model.Job`) - and that works as expected in most other places, I think. However, in this particular case we *know* it's never returning None, because the method is defined on the `JobWrapper` class which always has a `job_id` (it takes a Job arg in the constructor and assigns its `id` attr to `self.job_id`; and that Job should always belong to a session, I assume). So, I think, the type-ignore here is a legitimate way to help mypy make the right call based on the application logic that it cannot interpret on its own.

The other type-ignores (for return-value) vary. In some cases they point to a legitimate problem: a value can be None, which we don't always check before calling methods on that value (I don't remember which ones are like that). My intention was to (1) type-ignore those cases for now since those typing errors were not directly caused by a change introduced in the PR, but by the PR adding typing to the model (required by SA2.0), and that typing causing ripple effects throughout the code base: i.e., mypy type-checks a lot more now.  And then, after SA2.0 is part of dev, (2) address each newly added type-ignore - in which case I would either refactor to fix the problem (if one exists), and/or use your suggested approach for the case when `session.get` returns an unexpected type. Would that be OK?

nsoranzo replied:
The following should solve the problem with `None`:

```suggestion
        job_id = self.sa_session.get(Job, self.job_id)
        assert job_id
        return job_id
```

jdavcs replied:
@nsoranzo thanks, I'll apply this to the SA20 PR too.

nsoranzo commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1181,7 +1181,9 @@ def galaxy_url(self):
-        return self.sa_session.get(Job, self.job_id)
+        job_id = self.sa_session.get(Job, self.job_id)
+        assert job_id
+        return job_id
```
Just realised that this is actually a `job` not a `job_id`, so wrong variable name, sorry.

jdavcs replied:
I'll squash all the commits from the previous SA20 PR into one to avoid confusion (this one has to build on that one, which is not yet merged)

jdavcs committed:
Move user data access method from webapps to managers

jdavcs committed:
Initial data access testing infrastructure

jdavcs committed:
Add first library data access tests

jdavcs committed:
Add first user data access tests

jdavcs committed:
Start replacing test_galaxy_mapping w/data access tests

jdavcs committed:
More test-galaxy-mapping conversions

jdavcs committed:
Convert another test

jdavcs committed:
Convert test_ratings

jdavcs committed:
Refactor model fixtures

jdavcs committed:
Add Decimal to accpted types by util.nice_size()

jdavcs committed:
Drop test

This tested SQLAlchemy:
1. create: dataset, job
2. set: dataset.job = job
3. save, then load dataset by dataset.id; verify that dataset.job = job

jdavcs committed:
Drop test

Tested SQLAlchemy:
- create job
- set job.tool_id
- save, load job by id, verify job.tool_id

jdavcs committed:
Convert test_history_contents

jdavcs committed:
Convert test_current_galaxy_sesssion

jdavcs committed:
Convert hid tests

jdavcs committed:
Convert test_get_display_name

jdavcs committed:
Drop test_tags

Tests SQLAlchemy:

1. Create tag
2. Verify there's no FooTagAssociation with this tag
3. Add new FooTagAssociation to Foo and save to db
4. Verify there is one FooTagAssociation with this tag

jdavcs committed:
Drop incomplete test

No assert (only TODO); but would test SQLAlchemy

jdavcs committed:
Drop test_basic

Tests SQLAlchemy, user-history relationship mapping

jdavcs committed:
Convert test_metadata_spec

jdavcs committed:
Drop unused fixture arg

jdavcs committed:
Convert + improve test_job/task_metrics

jdavcs committed:
Drop test_tasks

Tests SQLAlchemy (create foo + set foo.bar + load foo + check bar)

jdavcs committed:
Use DeclarativeBase as base class

mypy bug workaround no longer needed as we are no longer specifying a metaclass

jdavcs committed:
Remove future, autocommit args from session and engine creation

jdavcs committed:
Fix table/__table__/HasTable for models

HasTable was a hack to accommodate declarative mapping without changing
thousands of lines that referred to the `table` attribute of model
instances.

The one type:ignore added to managers.datasets can be removed after we
map DatasetInstance models declaratively.

jdavcs committed:
Remove fix for model constructors (fixed in SA2.0)

jdavcs committed:
Add missing type hints to mapped_column in the model

jdavcs committed:
Fix bug: WorkflowInvocationMessage in not serializable

Uncovered by adding missing type hints to model

jdavcs committed:
Add type hint to 1 column_property to fix a new mypy error

jdavcs committed:
Fix bug: there's no user attr on WorkflowInvocation

jdavcs committed:
Fix typing in CustosAuthmzToken model

jdavcs committed:
Fix uuid types in the model

jdavcs committed:
Fix type error in workflow unit test

jdavcs committed:
Start trimming model definition syntax

Drop syntax that can be inferred from Mapped[] type hint:
- basic datatypes
- mapped_column() expression where the only argument is the datatype

Ref: https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#migrating-an-existing-mapping
